### PR TITLE
Check if remote hosts are reachable before making requests. Fixes gh-1051

### PR DESCRIPTION
**Pull request by rwaldron**
- Fails with error that indicates the cause

### CODE REVIEW

File: lib/crash-reporter.js [hunk 1]
```diff
@@ -149,30 +150,32 @@ CrashReporter.post = function(labels, crash, argv) {
...
           }
+        } catch (exception) {
+          reject(exception);
         }
-      } catch (exception) {
-        reject(exception);
-      }
-    });
+      });
+    }).catch(reject);
```
Comment (rwaldron) [inline lib/crash-reporter.js]: This is almost entirely whitespace change

File: lib/install/rust.js [hunk 2]
```diff
@@ -20,19 +20,17 @@ var tar = require('tar-fs');
...
-
-var SDK_URLS = {
-  macos: 'https://builds.tessel.io/t2/sdk/t2-sdk-macos-x86_64.tar.bz2',
-  linux: 'https://builds.tessel.io/t2/sdk/t2-sdk-linux-x86_64.tar.bz2',
+const SDK_URLS = {
+  macos: `https://${remote.BUILDS_HOSTNAME}/t2/sdk/t2-sdk-macos-x86_64.tar.bz2`,
+  linux: `https://${remote.BUILDS_HOSTNAME}/t2/sdk/t2-sdk-linux-x86_64.tar.bz2`,
 };
 
-var RUST_LIB_TGZ_URL = 'https://builds.tessel.io/t2/sdk/t2-rustlib-VERSION.tar.gz';
```
Comment (rwaldron) [inline lib/install/rust.js]: Created later, when `VERSION` is actually known

File: lib/tessel/deployment/javascript.js [hunk 3]
```diff
@@ -293,18 +294,20 @@ exportables.resolveBinaryModules = function(opts) {
...
+                })
+                .pipe(gunzip)
+                .pipe(extract)
+                .on('error', reject)
+                .on('end', () => {
+                  // Once complete, the locally cached binary can
+                  // be found in ~/.tessel/binaries
+                  resolve();
+                });
+            }).catch(reject);
```
Comment (rwaldron) [inline lib/tessel/deployment/javascript.js]: This is almost entirely whitespace change

File: lib/update-fetch.js [hunk 4]
```diff
@@ -38,33 +39,35 @@ var exportables = {
...
+
+          // Sort the builds by semver version in chronological order
+          builds.sort((a, b) => semver.compare(a.version, b.version));
+
+          return resolve(builds);
+        } else {
+          reject(outcome.reason);
+        }
+      });
+    }).catch(reject);
```
Comment (rwaldron) [inline lib/update-fetch.js]: This is almost entirely whitespace change

File: lib/update-fetch.js [hunk 5]
```diff
@@ -160,31 +163,34 @@ exportables.downloadTgz = function(tgzUrl, fileMap) {
...
-    });
+        // When we get incoming data, update the progress bar
+        res.on('data', (chunk) => {
+          bar.tick(chunk.length);
+        });
+
+        // unzip and extract the binary tarball
+        res.pipe(gunzip).pipe(extract);
+      });
+    }).catch(reject);
```
Comment (rwaldron) [inline lib/update-fetch.js]: This is almost entirely whitespace change

**Comment by rwaldron** [comment]
@HipsterBrown 

## Smoke Test

Disconnected internet: 

- `t2 version`
- `t2 run index.js` (with precompiled binary that isn't cached locally)

@tcr 

Connected internet: 

- Check that cargo rust compilation still works correcly

Disconnected internet: 

- Check that remote cc fails

**Comment by HipsterBrown** [comment]
I'll smoke test this tonight.

File: lib/install/rust.js [hunk 6]
```diff
@@ -20,19 +20,17 @@ var tar = require('tar-fs');
 
 // Internal
 var log = require('../log');
+var remote = require('../remote');
 
-var SDK_PATHS = {
+const SDK_PATHS = {
```
Comment (HipsterBrown) [inline lib/install/rust.js]: Are we standardizing on `const` now that Node 6 is LTS?

Comment (rwaldron) [inline lib/install/rust.js]: Yes, I'd say we should migrate in new code, or code that we're in touch with during patches.

**Pull request closed without merging**